src/raftstore/store/snap.rs

```diff
@@ -618,6 +665,987 @@ mod v1 {
+                Err(err) => return Err(err),
+            }
+        }
+    }
+
+    fn get_snapshot_meta(cf_files: &[CfFile]) -> RaftStoreResult<SnapshotMeta> {
+        let mut meta = Vec::with_capacity(cf_files.len());
+        for cf_file in cf_files {
+            if SNAPSHOT_CFS.iter().find(|&cf| cf_file.cf == *cf).is_none() {
+                return Err(box_err!("failed to encode invalid snapshot cf {}", cf_file.cf));
+            }
+
+            let mut cf_file_meta = SnapshotCFFile::new();
+            cf_file_meta.set_cf(cf_file.cf.to_owned());
+            cf_file_meta.set_size(cf_file.size);
+            cf_file_meta.set_checksum(cf_file.checksum);
+            meta.push(cf_file_meta);
+        }
+        let mut snapshot_meta = SnapshotMeta::new();
+        snapshot_meta.set_cf_files(RepeatedField::from_vec(meta));
+        Ok(snapshot_meta)
+    }
+
+    fn check_snapshot_meta(snapshot_meta: &SnapshotMeta) -> RaftStoreResult<Vec<CfName>> {
+        let mut res = Vec::with_capacity(SNAPSHOT_CFS.len());
+        for cf_file_meta in snapshot_meta.get_cf_files() {
+            let found = SNAPSHOT_CFS.iter().find(|&s| cf_file_meta.get_cf() == *s);
+            if found.is_none() {
+                return Err(box_err!("failed to decode invalid snapshot cf {}",
+                                    cf_file_meta.get_cf()));
+            }
+            let cf_name = found.unwrap();
+            if res.iter().any(|cf| cf_name == cf) {
+                return Err(box_err!("failed to decode duplicated snapshot cf {}", cf_name));
+            }
+            res.push(*cf_name);
+        }
+        if res.len() != SNAPSHOT_CFS.len() {
+            return Err(box_err!("invalid number of snapshot meta cf file: {}, expect: {}",
+                                res.len(),
+                                SNAPSHOT_CFS.len()));
+        }
+        Ok(res)
+    }
+
+    fn check_file_size_and_checksum(path: &PathBuf,
+                                    expect_size: u64,
+                                    expect_checksum: u32)
+                                    -> RaftStoreResult<()> {
+        let size = try!(get_file_size(path));
+        if size != expect_size {
+            return Err(box_err!("invalid size {} for snapshot cf file {}, expected {}",
+                                size,
+                                path.display(),
+                                expect_size));
+        }
+        let checksum = try!(calc_crc32(&path));
+        if checksum != expect_checksum {
+            return Err(box_err!("invalid checksum {} for snapshot cf file {}, expected {}",
+                                checksum,
+                                path.display(),
+                                expect_checksum));
+        }
+        Ok(())
+    }
+
+    #[derive(Default)]
+    struct CfFile {
+        pub cf: CfName,
+        pub path: PathBuf,
+        pub tmp_path: PathBuf,
+        pub sst_writer: Option<SstFileWriter>,
```

Comment:
It need three values here sst_writer, file, None. Then I think it wraps too much using Option<Either<SstFileWriter, File>>. It seems too verbose to fetch the inner value.